tripal/tripal
# Support Chado Flyway Migrations

# Tripal 4 Core Dev Task

### Issue #1876 


### Tripal Version: 4

## Description

This PR adds support to Tripal to manage chado migrations ourselves so that we don't add Flyway as an external migration and increase complexity for users. Users will still install Chado 1.3 as they do now. Once Chado 1.3 is installed, they can go to the Chado Schema Manager and apply any outstanding migrations as shown in the following screenshots.

<img width="1495" alt="ChadoSchemaManager" src="https://github.com/user-attachments/assets/d77dad16-efac-4c7e-b1d1-4e661baa74de">

<img width="1493" alt="ChadoSchemaManager-ApplyMigrations" src="https://github.com/user-attachments/assets/5cc1dd6b-2d39-4222-bae4-5be98f993b65">

## Tasks

- [x] Store the Chado migrations in a good place in Tripal
- [x] Use Yaml to provide nice metadata for each migration for use by Tripal
- [x] Add a button to the Chado Schema Manager form to manage migrations
- [x] Develop a framework to check for migrations to be applied and display them to the user
- [x] Use a tripal job to actually apply the migrations and then update the UI to show the user that they were applied.
- [x] Ensure that the chado.chadoprop table with the version number is updated to reflect the migration applied.

## Testing?




For testing, please focus your testing on the UI and migration application. Any issues with Tripal's functioning on higher versions are beyond the scope of this PR so please file and issue so they can be fixed in another PR :-)

1. In a TripalDocker based on this branch, go to Tripal > Data Storage > Chado > Chado Schemas. You should see something like the first screenshot above.
2. Click on "Apply Migrations" and you should see the second screenshot above.
3. Click "Apply Migrations" again and a Tripal job should be submitted.
4. When you run the Tripal job it should apply the migrations. Ensure the job runs without error and all the migrations are listed.
5. Now using check the chado schema to ensure that the migrations were actually applied. You can do this using `psql --user docker sitedb` where the password is the same as the username. Using `\d+ chado.TABLENAME` you can see the structure of the table. For example, `\d+ chado.project` should show the project now has a type_id and checking contact_project should show there is a type_id and rank.
6. Next, go back to the "Manage Chado Schemas" page described in step 1. The schema version should now match that of the latest migration (i.e. 1.3.3.013) and the updated column should be different from the created column.
<img width="1373" alt="Screenshot 2024-07-29 at 2 26 25 PM" src="https://github.com/user-attachments/assets/23e1a4fe-06a8-49b5-8620-3a95609ac81a">

7. Additionally, the "Apply Migrations" button should be replaced with a "View Migrations" button.
8. Click the "View Migrations" button and ensure they all say "Successful"
![image](https://github.com/user-attachments/assets/f35278ad-1149-444a-8e30-c4da653a36d1)
